tests/components/numato/test_numato.py

```diff
@@ -0,0 +1,325 @@
+                            "unit": "mocks",
+                        }
+                    },
+                },
+                "switches": {
+                    "invert_logic": False,
+                    "ports": {
+                        "5": "numato_switch_mock_port5",
+                        "6": "numato_switch_mock_port6",
+                    },
+                },
+            }
+        ],
+    }
+}
+
+
+def mockup_raise(*args, **kwargs):
+    """Mockup to replace regular functions for error injection."""
+    raise NumatoGpioError("Error mockup")
+
+
+def mockup_return(*args, **kwargs):
+    """Mockup to replace regular functions for error injection."""
+    return False
+
+
+@pytest.fixture
+def config():
+    """Provide a copy of the numato domain's test configuration.
+
+    This helps to quickly change certain aspects of the configuration scoped
+    to each individual test.
+    """
+    return deepcopy(NUMATO_CFG)
+
+
+@pytest.fixture
+def numato_fixture():
+    """Inject the numato mockup into numato homeassistant module."""
+    module_mock = copy(numato_mock.NumatoModuleMock)
+    numato.__dict__["gpio"] = module_mock
+    yield module_mock
+    module_mock.cleanup()
+
+
+@pytest.mark.asyncio
+async def test_setup_no_devices(hass, numato_fixture, monkeypatch):
+    """Test handling of an 'empty' discovery.
+
+    Platform setups are expected to return after handling errors locally
+    without raising.
+    """
+    monkeypatch.setattr(numato_fixture, "discover", mockup_return)
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    assert len(numato_fixture.devices) == 0
+
+
+@pytest.mark.asyncio
+async def test_fail_setup_raising_discovery(hass, numato_fixture, caplog, monkeypatch):
+    """Test handling of an exception during discovery.
+
+    Setup shall return False.
+    """
+    monkeypatch.setattr(numato_fixture, "discover", mockup_raise)
+    assert not await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()
+
+
+@pytest.mark.asyncio
+async def test_regular_hass_operations(hass, numato_fixture):
+    """Test regular operations from within Home Assistant."""
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()  # wait until services are registered
+    await async_turn_on(hass, "switch.numato_switch_mock_port5")
+    assert numato_fixture.devices[0].values[5] == 1
+    await async_turn_on(hass, "switch.numato_switch_mock_port6")
+    assert numato_fixture.devices[0].values[6] == 1
+    await async_turn_off(hass, "switch.numato_switch_mock_port5")
+    assert numato_fixture.devices[0].values[5] == 0
+    await async_turn_off(hass, "switch.numato_switch_mock_port6")
+    assert numato_fixture.devices[0].values[6] == 0
+
+    numato_fixture.devices[0].mockup_inject_notification(2, 1)
+    await hass.async_block_till_done()
+    state = hass.states.get("binary_sensor.numato_binary_sensor_mock_port2")
```

Comment:
We want to assert the state of the switch too, after each service call. And the binary sensor tests should be a separate test and module.

Reply:
Done.